src/error-table.rkt

```diff
@@ -1,20 +1,460 @@
 #lang racket
 
-(require "points.rkt" "syntax/types.rkt" "core/localize.rkt" "timeline.rkt"
-         "common.rkt")
+(require racket/set math/bigfloat)
+(require "points.rkt" "syntax/types.rkt" "core/localize.rkt" "common.rkt"
+         "ground-truth.rkt" "syntax/sugar.rkt")
 
-(provide group-errors)
+(provide actual-errors predicted-errors)
 
-(define (group-errors expr pcontext)  
+(define (actual-errors expr pcontext)
   (match-define (cons subexprs pt-errorss)
     (flip-lists
-     (hash->list (car (compute-local-errors (list expr) (*context*))))))
-
+     (hash->list (car (compute-local-errors (list expr)
+                                            (*context*))))))
+  
   (define pt-worst-subexprs
-    (for/list ([pt-errors (in-list pt-errorss)])
-      (match-define (cons worst-subexpr pt-max-error)
-        (argmax cdr (map cons subexprs pt-errors)))
-      (if (> pt-max-error 2) worst-subexpr #f)))
+    (foldr append
+           '()
+           (for/list ([pt-errors (in-list pt-errorss)]
+                      [(pt _) (in-pcontext pcontext)])
+             (define sub-error (map cons subexprs pt-errors))
+             (define filtered-sub-error
+               (filter (lambda (p) (> (cdr p) 16)) sub-error))
+             (define mapped-sub-error
+               (map (lambda (p) (cons (car p) pt))
+                    filtered-sub-error))
+             (if (empty? mapped-sub-error)
+                 (list (cons #f pt))
+                 mapped-sub-error))))
+
+  (for/hash ([group (in-list (group-by car pt-worst-subexprs))])
+    (let ([key (caar group)])
+      (values key (map cdr group)))))
+
+(define (same-sign? a b)
+  (or (and (positive? a) (positive? b))
+      (and (negative? a) (negative? b))))
+ 
+(define (predicted-errors expr ctx pctx)
+  
+  (define old-precision (bf-precision))
+  (bf-precision 1000)
+  (define cond-thres (bf 100))
+  
+  (define subexprs
+    (all-subexpressions-rev expr (context-repr ctx)))
+  (define subexprs-list (map car subexprs))
+  (define spec-list (map prog->spec subexprs-list))
+ 
+  (define ctx-list
+    (for/list ([subexpr (in-list subexprs)])
+      (struct-copy context ctx [repr (cdr subexpr)])))
+
+  (define repr-hash
+    (make-immutable-hash (map cons
+                              subexprs-list
+                              (map context-repr ctx-list))))
+ 
+  (define subexprs-fn (eval-progs-real spec-list ctx-list))
+ 
+  (define error-count-hash
+    (make-hash (map (lambda (x) (cons x '()))
+                    (cons #f subexprs-list))))
```

Comment:
I still think i'd have to prepopulate the hash, because let's say ground truth has 100 predictions and we predict 0. Then there is no entry for this subexpression in predicted. The same thing can happen with no entry for that subexpression in ground truth. So predictions and ground truth are not perfect subsets of each other.